src/multicall.provider.ts

```diff
@@ -0,0 +1,212 @@
+    multicallAddress: string,
+    provider: EthersProvider,
+    blockTag?: number | string
+  ): Promise<T> {
+    const Multicall2 = new Contract(multicallAddress, MULTICALL2_ABI, provider);
+    let results = [];
+    let success = true;
+    // split calls into batches of size `this._batchSize`.
+    const subCalls = this.batchArray(calls, this._batchSize);
+
+    try {
+      for (let callsBatch of subCalls) {
+        const requests = callsBatch.map((call) => ({
+          target: call.contract.address,
+          callData: Abi.encode(call.name, call.inputs, call.params),
+        }));
+
+        const response = await Multicall2.callStatic.tryAggregate(true, requests, { blockTag });
+
+        for (let i = 0; i < callsBatch.length; i++) {
+          const outputs = callsBatch[i].outputs;
+          const result = response[i];
+          const params = Abi.decode(outputs, result.returnData);
+          results.push(outputs.length === 1 ? params[0] : params);
+        }
+      }
+    } catch (e) {
+      // One or more of the calls failed.
+      success = false;
+      results = [];
+    }
+
+    return [success, results] as T;
+  }
+
+  // internal function called by `tryAll`, `groupTryAll`.
+  private async _tryAll<T extends any[] = any[]>(
+    calls: ContractCall[],
+    multicallAddress: string,
+    provider: EthersProvider,
+    blockTag?: number | string
+  ): Promise<T> {
+    const Multicall2 = new Contract(multicallAddress, MULTICALL2_ABI, provider);
+    const results = [];
+
+    // split calls into batches of size `this._batchSize`.
+    const subCalls = this.batchArray(calls, this._batchSize);
+
+    for (let callsBatch of subCalls) {
+      const requests = callsBatch.map((call) => ({
+        target: call.contract.address,
+        callData: Abi.encode(call.name, call.inputs, call.params),
+      }));
+
+      const response = await Multicall2.callStatic.tryAggregate(false, requests, { blockTag });
+
+      for (let i = 0; i < callsBatch.length; i++) {
+        const outputs = callsBatch[i].outputs;
+        const result = response[i];
+
+        if (result.success) {
+          const params = Abi.decode(outputs, result.returnData);
+          results.push({
+            success: result.success,
+            returnData: outputs.length === 1 ? params[0] : params,
+          });
+        } else {
+          results.push({
+            success: result.success,
+            returnData: "0x",
+          });
+        }
+      }
+    }
+
+    return results as T;
+  }
+
+  // helper function to split the calls array into sub-arrays of size n.
+  private batchArray(calls: any[], n: number): any[][] {
+    const subArrays = [];
+
+    if (n == 0) return [[]];
+    let i;
+    for (i = 0; i < calls.length - n; i = i + n) {
+      subArrays.push(calls.slice(i, i + n));
+    }
+    if (i < calls.length) subArrays.push(calls.slice(i));
+    return subArrays;
+  }
+}
+
+export { MulticallContract, MulticallProvider };
```

Comment:
All the changes make sense to me. Thanks for taking time to update and improve the code! 

I'm only a bit hesitant about calling `_tryAll` instead of `tryAggregate` in `_all` func. Both the old and updated versions will deliver the same results. But what made me go with using `tryAggregate` is the following:  

When calling `_tryAll` the multicall contract will continue executing all the calls even if one of them failed, this is unneeded and will imply additional execution time ( especially if the calls length is big and the first or second call fails). However, using `tryAggregate(requireSuccess = true)` will cause the tx to revert sooner. wdyt?

Same note for calling `groupTryAll` inside `groupAll`.

Reply:
Hmm, about `_tryAll` you're right it will take additional time for execution. My opinion about this is that the additional execution time should really not be significant when compared with the actual network communication time. But if there's anything that points towards this not being the case or you still think it's better to make a call with `requireSuccess` set, I'm fine with both approaches, no problem.

About `groupAll`, I do think using `_tryAll` internally is better. The reason for that is what I think would be the expected behavior for grouped calls is "do this kind of call for each group" - basically, the expected behavior imo for `_tryAll([a, b, c])` is returning the same output as `[all(a), all(b), all(c)]`, not `split(all([...a, ...b, ...c]))`. What do you think about this?

Reply:
Yes the time shouldn't be significant but I don't see a reason why we need to switch to using `_tryAll` so I still prefer to keep `tryAggregate`. 

About using `_tryAll` inside `groupAll`, my understanding of the grouped calls was to do the same as `all`/ `tryAll` for all the calls flattened **with preserving the input structure**. That's why I used `_all`. 

 I mean, if we use `_tryAll` for `groupAll` what's the point of calling the rest of groups if the first group failed? If we want to get the result of each group we'll use `groupTryAll` instead. And If we're concerned about not making all calls at once, we have the `batchSize` for that. wdyt?

Reply:
Alright, no problem here.

> About using `_tryAll` inside `groupAll`, my understanding of the grouped calls was to do the same as `all`/ `tryAll` for all the calls flattened **with preserving the input structure**. That's why I used `_all`. 
Yes, I mean, this is really what I meant in the beginning.

>  I mean, if we use `_tryAll` for `groupAll` what's the point of calling the rest of groups if the first group failed? If we want to get the result of each group we'll use `groupTryAll` instead. And If we're concerned about not making all calls at once, we have the `batchSize` for that. wdyt? 
Yes, what I thought `groupAll` and `groupTryAll` could do is basically just instead of making multiple multicalls in a for-each or so, you just pass those groups and the calls would be made in the same manner. In `tryAll`, this would mean what I said in the comment above.
The thing with using `groupTryAll` for cases some of the calls can fail is that each individual call would have a success flag, as opposed to each group having a success flag. Then it'd be necessary to examine those individually and get the return data. This works too, but I don't see a clear use-case for having `groupAll` fail when any of the calls fail, **if** we consider that the groups are independent from each other. Ofc, we have `batchSize`, but if multiple `all`s are done in a loop to avoid doing a `tryAll`, then the calls aren't really batched.

So that's the design choice, I think. Should we consider the groups independent from one another or should we consider that the goal is only keeping the input structure?

Reply:
My thoughts are:
1. I prefer to use `requireSuccess = true` accordingly.
2. I think that is better to us the `group` version of the methods just to preserve the input structure.

About 2, I think that was the initial goal, because adding support for considering the groups independent seems like something not worth to do (for me) because it is something easily achievable with a `map` on the user side.

Please notice all the improvements in the Franco suggestion and make sure of persist them.

Reply:
:+1: About the latest messages, all good here. 

Also, @e1Ru1o and @yasmine9, wdyt about moving this PR to be based on the `feat/v3` branch? Looks like a good way to introduce it together with some more new features.

Reply:
Yes, I think it's a good idea to move this to the V3 👍. 


So, I updated the class based on your comments and suggestions. I modified `groupAll` and `_all` implementations to keep using the old approach. I updated `GroupAllResult` type too. The rest is basically from Franco's suggestions.